### PR TITLE
Implementation of conditional dialogue and action-dialogue integrations; Wrote dialogue unit tests

### PR DESCRIPTION
**Pull request by jeyeong**
Introduced the following changes to the dialogue module.

**Conditional Dialogue**
- Allow dialogue designers to designate certain edges (i.e. dialogue options) as "conditional," that is, only accessible if the player has fulfilled certain conditions.
- The implementation makes use of conditions from game-state/conditions.h.
- To make start_conversation() and run_conversation_step() work with conditional dialogue, some revisions had to be made: (1) Implemented a function to check for which edges are "available" upon reaching the node. (2) Generating the option numbers in the return string based on these "available" nodes.
- The edge and edge_list struct also had to be revised.
- Multiple conditions are possible.

**Action Integrations**
- Allow nodes to have associated actions, for example, giving the player an item, or starting a quest.
- This was achieved by implementing a node_action struct (a linked list) and an associated do_node_actions() function. Upon reaching each node, do_node_actions() is called, executing each action.
- Multiple actions per node are possible.

**Unit Tests** (in npc/test_dialogue.c)
- Tests for the basic functions, such as the init, new, free functions, as well as add_node() and add_edge().
- Tests for dialogue execution, i.e. start_conversation() and run_conversation_step().
- Tests for conditional dialogue.
- Tests for actions.

A working example of conditional dialogue + actions can be found in **src/npc/examples/dialogue_example.c.**

### CODE REVIEW

File: src/npc/src/dialogue.c [hunk 1]
```diff
@@ -185,35 +261,127 @@ char *create_return_string(convo_t *c, int is_leaf)
...
+    if (input > c->cur_node->num_available_edges)
+        input = c->cur_node->num_available_edges;
+
+    char *ret_str;
     edge_list_t *cur_edge;
-    int i;
 
+    // Step 1: Traverse to the player's selected node
     cur_edge = c->cur_node->edges;
-    for (i = 1; i < input && i < c->cur_node->num_edges; i++) {
```
Comment (pkasner) [inline src/npc/src/dialogue.c:206]: The logic in this step is a little difficult to follow. Not sure if that's something to clarify with more comments or just me being slow to process how the while conditions interact with the if statement. Perhaps specify that the input value associated with an edge is equal to its order in the linked list ignoring any unavailable edges before it.

File: tests/npc/test_dialogue.c [hunk 2]
```diff
@@ -302,3 +307,368 @@ Test(dialogue, add_bidirectional_edges)
...
+
+    cr_assert_eq(rc, 0, "Return Code was set to %d when it should have been 0, "
+                 "indicating that the conversation has not ended", rc);
+    cr_assert_eq(strcmp(ret_str, expected), 0,
+                 "Expected:\n%s for the return string of start_conversation "
+                 "but start_conversation returned:\n%s", expected, ret_str);
+}
+
+/* 2 normal edges and 2 conditional edges, one satisfied and one unsatisfied,
+   results in 2 available edges */
```
Comment (pkasner) [inline tests/npc/test_dialogue.c:548]: Some of these tests are probably redundant/don't seem to add additional coverage. No reason to delete them, but something to be aware of for the future.

File: include/npc/dialogue.h [hunk 3]
```diff
@@ -127,29 +168,67 @@ int add_edge(convo_t *c, char *quip, char *from_id, char *to_id);
...
+ * Parameters:
+ *  - c: pointer to a convo
+ *  - node_id: ID of the target node
+ *  - quest_id: ID of the quest
+ *
+ * Returns:
+ *  - SUCCESS on success, FAILURE if an error occurs
+ *  - Possible errors: (1) node matching node_id could not be found;
+ *    (2) the node already has an action;
+ */
```
[the remark this answers is not in the document]
Comment (pkasner) [reply include/npc/dialogue.h:216]: Because the actual initiating of quests is left as to do, so this should be fine for now.

**Review by pkasner — CHANGES_REQUESTED** [review]
For the most part looks good. Obviously the tests need to pass, so those will need to be figured out. Most important change is probably making sure item possession is handled gracefully for the npc and player if there is a transfer during dialogue. Also this perhaps would have been better as two pull requests, one for conditional dialogue and one for actions.

File: include/npc/dialogue.h [hunk 4]
```diff
@@ -12,6 +15,33 @@
...
+ * dialogue. NOTE: This is a linked list, for multiple actions.
+ * 
+ * Includes:
+ *  - action: the type of action
+ *  - action_id: ID associated with the action (e.g. quest ID), if any
+ *  - next, prev: next and previous list elements
+ */
+typedef struct node_action {
+    node_action_type action_type;
+    char *action_id;
```
[the remark this answers is not in the document]
Comment (jeyeong) [reply include/npc/dialogue.h:35]: Yes, that's correct. Action_type identifies the action being taken. Though it seems like that isn't so clear.

File: include/npc/dialogue.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (jeyeong) [reply include/npc/dialogue.h:216]: Ah, I see. Technically, it isn't completely out of the question to convert a string into a long int, and vice versa.

File: src/npc/src/dialogue.c [hunk 5]
```diff
@@ -121,31 +115,111 @@ int add_edge(convo_t *c, char *quip, char *from_id, char *to_id)
...
+
+    while (action != NULL) {
+
+        switch(action->action_type) {
+
+            case GAIN_ITEM: ;
+                item_t *item = get_item_in_hash(game->all_items,
+                                                action->action_id);
+                if (item == NULL) return FAILURE;
+                int rc = add_item_to_player(game->curr_player, item);
```
[the remark this answers is not in the document]
Comment (jeyeong) [reply src/npc/src/dialogue.c:139]: That's actually a good point. One thing we would have to provide if we want to remove the item from the NPC's inventory is a pointer to the NPC we are currently speaking to. Which we may have to think about.

[the remark this answers is not in the document]
Comment (jeyeong) [reply src/npc/src/dialogue.c:139]: This current code itself, however, fails generally due to the item bug. I was considering fixing that error before moving on to a more proper implementation.

File: src/npc/src/dialogue.c (hunk 1 above)
Comment (jeyeong) [reply src/npc/src/dialogue.c:206]: How did that get there? Was this from an old commit? The original idea of "i < c->cur_node->num_edges" was to handle the event where input is > than the number of available options. But I've since changed that implementation!

File: src/npc/src/dialogue.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (pkasner) [reply src/npc/src/dialogue.c:139]: With regard to keeping track of the npc we're talking to, I notice this function takes the game struct as a parameter, which means it can access the game's mode. In conversation mode, the mode structs mode_ctx field contains the npc's npc_id.

File: src/npc/src/dialogue.c (hunk 1 above)
Comment (pkasner) [reply src/npc/src/dialogue.c:206]: I think I clicked the wrong line with this comment. Check the code immediately underneath maybe?

File: tests/npc/test_dialogue.c (hunk 2 above)
Comment (jeyeong) [reply tests/npc/test_dialogue.c:548]: Thanks for pointing that out. My thought process for the conditional dialogue tests were as follows:
1. Conditional edge, not passed
2. Conditional edge, passed
3. 1 Unconditional edge + 1 Conditional edge, passed (testing for interaction between conditional, unconditional edges)
4. 2 Unconditional edges + 2 Conditional edges, 1 pass, 1 fail (testing for interaction between the two conditional edges)

Are you saying including test #4 alone would perhaps have sufficed?

File: include/npc/dialogue.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (jeyeong) [reply include/npc/dialogue.h:35]: I renamed action_type to simply action

File: src/npc/src/dialogue.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (jeyeong) [reply src/npc/src/dialogue.c:139]: Ah, that would be convenient. Can you point me to the function that returns the npc struct using the npc_id?

File: src/npc/src/dialogue.c (hunk 1 above)
Comment (jeyeong) [reply src/npc/src/dialogue.c:206]: Ah, I see what you mean by hard to follow. This was the logic I needed to implement to make sure we end up at the correct edge. Perhaps I should state that explicitly in a comment.

File: tests/npc/test_dialogue.c (hunk 2 above)
Comment (pkasner) [reply tests/npc/test_dialogue.c:548]: I think 1 and 4 probably would have sufficed. 1 still seems like a useful edge case (as it returns 1 instead of 0 like the others).

File: src/npc/src/dialogue.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (pkasner) [reply src/npc/src/dialogue.c:139]: There currently is no such function, but it can be done with the HASH_FIND uthash macro. However, this also assumes that the npc/mode_operation branch is merged, which adds the all_npcs field as a hast table to chiventure's game struct. I do think I will probably add a get_npc function or something of the sort to that branch in the npc module.

[the remark this answers is not in the document]
Comment (pkasner) [reply src/npc/src/dialogue.c:139]: The get_npc() function in game.h in the npc/mode_operation branch now does that job.

File: tests/npc/test_dialogue.c (hunk 2 above)
Comment (jeyeong) [reply tests/npc/test_dialogue.c]: Got it, that makes sense. I've removed the redundant 2 & 3.

File: src/npc/src/dialogue.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (jeyeong) [reply src/npc/src/dialogue.c:139]: Great!

**Comment by pkasner** [comment]
This looks good! I'm tentatively approving this, but I think we should still wait to merge this until after the npc/mode_operation branch can be merged. That way we can make sure conversation mode still works and we can do some integration testing.

**Review by pkasner — APPROVED** [review]
My above see my above comment that should've been attached to this review.

**Comment by jeyeong** [comment]
After @pkasner 's pull request got merged, the code no longer compiled. I've made the changes so that it compiles now.

**Comment by jeyeong** [comment]
I remodelled the GIVE_ITEM and TAKE_ITEM actions so that they interact with NPCs rather than the game. Also, since npc_new() was redefined in a recent update, I updated the example files so that they now work.

File: src/npc/src/dialogue.c [hunk 6]
```diff
@@ -121,58 +115,157 @@ int add_edge(convo_t *c, char *quip, char *from_id, char *to_id)
...
+    int len_lst[n->num_available_edges + 2];
+    edge_list_t *cur_edge = n->edges;
     int i = 0;
+    int option_number = 1;
 
     // Compute buffer length
-    totlen += (len_lst[i++] = strlen(c->cur_node->npc_dialogue));
-    totlen += 2;
+    totlen += (len_lst[i++] = strlen(n->npc_dialogue));
+    totlen += 1;
```
Comment (lilyehsani) [inline src/npc/src/dialogue.c:233]: Was the change from 2 to 1 here intentional?

File: src/npc/src/dialogue.c [hunk 7]
```diff
@@ -185,35 +278,143 @@ char *create_return_string(convo_t *c, int is_leaf)
...
+/* See dialogue.h */
+int add_start_quest(convo_t *c, char *node_id, char *quest_id)
+{
+    assert(quest_id != NULL);
+    
+    node_t *n;
+    if ((n = get_node(c->all_nodes, node_id)) == NULL) return FAILURE;
+
+    return add_action_to_node(n, START_QUEST, quest_id);
 }
```
Comment (lilyehsani) [inline src/npc/src/dialogue.c:419]: Again, should probably write a to-do for add_start_battle

File: src/npc/src/dialogue.c [hunk 8]
```diff
@@ -270,39 +469,27 @@ int edge_free(edge_t *e)
...
+    n->actions = NULL;
 
     return SUCCESS;
 }
 
 /* See dialogue.h */
 node_t *node_new(char *node_id, char *npc_dialogue)
 {
-    assert(strlen(node_id) < MAX_NODE_ID_LEN);
-    assert(strlen(npc_dialogue) < MAX_DIA_LEN);
```
Comment (lilyehsani) [inline src/npc/src/dialogue.c:298]: Why delete these?

File: src/npc/src/dialogue.c [hunk 9]
```diff
@@ -389,3 +579,33 @@ int free_node_list(node_list_t *n_lst, bool free_nodes)
 
     return SUCCESS;
 }
+
+/* See dialogue.h */
+node_action_t *node_action_new(node_action_type action, char *action_id)
```
Comment (lilyehsani) [inline src/npc/src/dialogue.c:584]: Should write corresponding init/free functions for this.

**Review by lilyehsani — CHANGES_REQUESTED** [review]
Really great job overall! Added a few comments, many of which are just questions.

File: include/npc/dialogue.h [hunk 10]
```diff
@@ -22,39 +53,45 @@ typedef struct node node_t;
...
+/* A doubly-linked list containing edges and their "availabilities."
  *
  * Includes:
- *  - option_number: an edge's number label
+ *  - availablility: -1 = DISABLED, 0 = UNAVAILABLE, 1 = AVAILABLE (see above)
  *  - edge: the edge
  *  - next, prev: next and previous list elements
  */
 typedef struct edge_list {
-    int option_number;
```
[the remark this answers is not in the document]
Comment (jeyeong) [reply include/npc/dialogue.h:40]: No, I've updated the create_return_string() function so it works independent of `option_number`.

File: include/npc/dialogue.h [hunk 11]
```diff
@@ -127,29 +167,78 @@ int add_edge(convo_t *c, char *quip, char *from_id, char *to_id);
...
+ * Parameters:
+ *  - c: pointer to a convo
+ *  - node_id: ID of the target node
+ *  - quest_id: ID of the quest
+ *
+ * Returns:
+ *  - SUCCESS on success, FAILURE if an error occurs
+ *  - Possible errors: (1) node matching node_id could not be found;
+ */
+int add_start_quest(convo_t *c, char *node_id, char *quest_id);
```
[the remark this answers is not in the document]
Comment (jeyeong) [reply include/npc/dialogue.h:241]: I see. Would it be better to: (1) convert to long int into a string (using sprintf), or (2) create a union struct that could hold different types of IDs?

File: include/npc/dialogue.h [hunk 12]
```diff
@@ -127,29 +167,78 @@ int add_edge(convo_t *c, char *quip, char *from_id, char *to_id);
...
```
[the remark this answers is not in the document]
Comment (jeyeong) [reply include/npc/dialogue.h:243]: Done

File: include/npc/dialogue.h (hunk 11 above)
[the remark this answers is not in the document]
Comment (jeyeong) [reply include/npc/dialogue.h:241]: Actually, I'll just leave it as a to-do for now, since I don't believe the quest team has fully implemented quests into Chiventure

[the remark this answers is not in the document]
Comment (jeyeong) [reply include/npc/dialogue.h:241]: Actually, I'll just leave it as a to-do for now, since I don't believe we are able to integrate quests with dialogue at this point in time

File: src/npc/examples/npc_example.c [hunk 13]
```diff
@@ -30,7 +30,7 @@ convo_t *create_sample_convo()
     // Starting to build the conversation structure
     convo_t *c = convo_new();
```
[the remark this answers is not in the document]
Comment (jeyeong) [reply src/npc/examples/npc_example.c:31]: Done

[the remark this answers is not in the document]
Comment (jeyeong) [reply src/npc/examples/npc_example.c:31]: I only removed the commented code, instead of deleting the file because I think this was one of the files Prof. Sotomayor wanted students to refer to to get an idea of how monkeypatching works.

File: src/npc/src/dialogue.c (hunk 6 above)
Comment (jeyeong) [reply src/npc/src/dialogue.c:233]: It was, the reason is because previously, I had wanted two newlines to be printed. But I realized that the CLI doesn't handle multiple newlines and only prints 1, so I changed it to only have 1 newline.

File: src/npc/src/dialogue.c (hunk 7 above)
Comment (jeyeong) [reply src/npc/src/dialogue.c:419]: Done

File: src/npc/src/dialogue.c (hunk 8 above)
Comment (jeyeong) [reply src/npc/src/dialogue.c:298]: I realized that I was doing assertion testing twice: once in new() and once in init(). So I removed the assertions in new() and kept them in init()

File: src/npc/src/dialogue.c (hunk 9 above)
Comment (jeyeong) [reply src/npc/src/dialogue.c]: Done

Comment (jeyeong) [reply src/npc/src/dialogue.c]: The free function was free_action_list() but I guess that wasn't very clear. I've renamed it to free_node_actions()

File: src/npc/src/dialogue.c (hunk 6 above)
Comment (lilyehsani) [reply src/npc/src/dialogue.c:233]: Sounds good

File: src/npc/src/dialogue.c (hunk 8 above)
Comment (lilyehsani) [reply src/npc/src/dialogue.c:298]: Ah I see

File: include/npc/dialogue.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (lilyehsani) [reply include/npc/dialogue.h:40]: Gotcha

File: include/npc/dialogue.h (hunk 11 above)
[the remark this answers is not in the document]
Comment (lilyehsani) [reply include/npc/dialogue.h]: Sounds good

File: src/npc/examples/npc_example.c (hunk 13 above)
[the remark this answers is not in the document]
Comment (lilyehsani) [reply src/npc/examples/npc_example.c:31]: Awesome!

**Review by lilyehsani — APPROVED** [review]
LGTM! Approved.

**Comment by jeyeong** [comment]
Summary of changes made:
- Added conditional dialogue functionality to dialogues.
- Added actions to dialogue (e.g. you can gain an item from a dialogue node).
- Wrote dialogue unit tests.

Next steps:
- Complete the load_npc module, so that NPCs can be loaded from WDL files.
- Write a working game demo featuring NPCs and dialogue.

**Comment by lilyehsani** [comment]
Issue Score: ✔️+

Comments:
Great work overall! I have one comment:
- This PR represents a large amount of work that could've been integrated into `dev` through smaller PRs earlier in the quarter. Remember that you should not treat PRs as a singular event to merge all or most of your work into `dev`. Integrating incrementally can help identify integration issues early, as well as ensure that you are not rushing to merge all your code towards the end of a sprint. For example, the conditional dialogue, dialogue actions, and unit tests could have been separated into 3 or some combination of 2 PRs.
